File: 0055-jump-game/0055-jump-game.py
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        n = len(nums)
        dp = [0] * n
        dp[n - 1] = 1
        
        for i in range(n - 2, -1, -1):
            furthest = min(i + nums[i], n - 1)
            for nxt in range(i + 1, furthest + 1):
                if dp[nxt] == 1:
                    dp[i] = 1
                    break
        
        return dp[0] == 1
        
        

File: 0055-jump-game/0055-jump-game.cpp
class Solution {
public:
    bool canJump(vector<int>& nums) {
        int n = nums.size();
        vector<int> dp(n);
        dp[n - 1] = 1;
        
        for (int i = n - 2; i >= 0; --i){
            int furthest = min(n - 1, nums[i] + i);
            for (int nxt = i + 1; nxt <= furthest; ++nxt){
                if (dp[nxt] == 1){
                    dp[i] = 1;
                    break;
                }
            }
        }
        
        return dp[0] == 1;
    }
};